contents/ai/ml.md

```diff
@@ -0,0 +1,149 @@
+
+In supervised learning, the data used for training is accompanied with the desired output labels. Learning is 'supervised' as the algorithm is told whether it has made a correct prediction. In unsupervised learning, no labels are given to the algorithm.
+
+Popular resources for learning ML algorithms:
+- Machine Learning: [Stanford CS229](cs229.stanford.edu/)
+- Convolutional Neural Networks for Visual Recognition: [Stanford CS231n](http://cs231n.stanford.edu/)
+- Natural Language Processing with Deep Leraning: [Stanford CS224n](cs224n.stanford.edu/)
+- A practical approach towards neural networks: [University of San Francisco, FastAI](http://www.fast.ai/)
+
+### Types of Data
+Data is broadly split into 2 categories, structured and unstructured. 
+
+Structured data refers to data in a standardized format with categories and values. Structured data tends to be formatted like the data in relational databases. A good example would be sales records, with fields like date, quantity sold, location and such. Unstructured data refers to data without a predefined data model, such as images, audio and textual data.
+
+The distinction between the types of data is important as different ML algorithms perform better on different types of data. In particular, recent advances in neural networks has shown that they perform extremely well on unstructured data.
+
+## Machine Learning Workflow (Prototyping to Production)
+Bringing a machine learning algorithm in production requires a workflow which differs greatly from that of software engineering. This is due to a focus on prototyping. Prototyping is required as machine learning algorithms vary in efficacy when used in different domains. The quantity, quality and type of data differs greatly as well, sometimes containing errors such as corruption. Machine learning algorithms contain various parameters that can be tuned to improve performance.
+
+The prototyping phase in ML is usually done on [Jupyter notebooks](http://jupyter.org/), or [RStudio](https://www.rstudio.com/products/rstudio/). The highlight of such software is their strong support for visualization and quick code iterations. Visualizations are essential as they help practitioners analyze data and share findings with coworkers. Code style and clarity is less of a priority at this stage. 
+
+A machine learning algorithm is also refered to as a model after the algorithm is provided data to learn from.
+
+Prototyping can be broken down into the following phases:  
+1) Basic data preprocessing  
+2) Partitioning of data   
+3) Model training, evaluation and data analysis  
+4) Repeating step 3  
+
+### Basic Data Preprocessing
+Basic data preprocessing is required to remove unwanted noise from the data. Examples of unwanted noise are repeated content and corrupted images. The unwanted content is usually discarded. Training a model on such data hinders its ability to learn the actual patterns in uncorrupted data. 
```

Comment:
Unwanted content are usually **manually** discarded? Could mention it could be tedious. Or whether there are other better ways.